File: backend/app/domain/workout.py
from dataclasses import dataclass
from enum import StrEnum
from typing import NewType
from datetime import date
from pydantic import Field
from ulid import ULID

from app.domain.entity import EntityModel
from app.domain.exercise import ExerciseSlug
from app.domain.track import TrackId


# todo: Нужно заменить на uuid
WorkoutId = NewType("WorkoutId", ULID)


class WorkoutProtocolType(StrEnum):
    DEFAULT = "DEFAULT"


class WorkoutProtocol(EntityModel):
    """
    Протокол тренировки
    """

    type: WorkoutProtocolType = Field(default=WorkoutProtocolType.DEFAULT)
    title: str
    description: str = ""


class WorkoutExercise(EntityModel):
    """
    Упражнение в тренировке
    """

    exercise_slug: ExerciseSlug

    # Количество повторов
    repetitions: int | None = None
    # Текстовое описание повторов
    repetitions_text: str = ""

    # Вес (кг)
    weight: int | None = None
    # Текстовое описание веса
    weight_text: str = ""


class WorkoutSection(EntityModel):
    """
    Часть тренировки
    """

    title: str
    protocol: WorkoutProtocol
    exercises: list[WorkoutExercise] = Field(min_length=1)


class Workout(EntityModel):
    """
    Модель тренировки

    Может состоять из нескольких частей
    """

    id: WorkoutId = Field(default_factory=ULID)

    date: date
    track_id: TrackId

    # Упорядоченный список частей
    sections: list[WorkoutSection] = Field(min_length=1)


@dataclass(frozen=True)
class WrokoutCriteria:
    limit: int


class WorkoutRepository:
    def add(self, workout: Workout) -> Workout:
        pass

    def get_by_id(self, workout_id: WorkoutId) -> Workout | None:
        pass

File: backend/app/store/workout.go
package store

import (
	"github.com/oklog/ulid/v2"
)

type WorkoutID string
type WorkoutProtocolType string

const (
	WorkoutProtocolTypeDefault WorkoutProtocolType = "default"
)

type Workout struct {
	ID WorkoutID

	Date    string
	TrackID TrackID

	Sections []WorkoutSection
}

type WorkoutSection struct {
	Title     string
	Protocol  WorkoutProtocol
	Exercises []WorkoutExercise
}

type WorkoutProtocol struct {
	Type        WorkoutProtocolType
	Title       string
	Description string
}

type WorkoutExercise struct {
	ExerciseSlug ExerciseSlug

	Repetitions     int
	RepetitionsText string

	Weight     int
	WeightText string

	Description string
}

func CreateWorkoutId() WorkoutID {
	id := ulid.Make()
	return WorkoutID(id.String())
}

type WorkoutStore interface {
	Store

	Create(workout *Workout) (*Workout, error)
	Update(workout *Workout) (*Workout, error)
	Get(id WorkoutID) (*Workout, error)
	Find(criteria *WorkoutFindCriteria) ([]*Workout, error)
}

type WorkoutFindCriteria struct {
	TrackID TrackID
	Limit   int
}

func ExtractSlugsFromWorkouts(workouts []*Workout) []ExerciseSlug {
	slugs := make(map[ExerciseSlug]bool)

	for _, workout := range workouts {
		for _, section := range workout.Sections {
			for _, exercise := range section.Exercises {
				slugs[exercise.ExerciseSlug] = true
			}
		}
	}

	slugsList := make([]ExerciseSlug, 0)
	for slug := range slugs {
		slugsList = append(slugsList, slug)
	}

	return slugsList
}

func ClearUnknownExercisesSlugs(workout *Workout, knownSlugs map[ExerciseSlug]bool) {
	for i := range workout.Sections {
		for j := range workout.Sections[i].Exercises {
			if _, ok := knownSlugs[workout.Sections[i].Exercises[j].ExerciseSlug]; !ok {
				workout.Sections[i].Exercises[j].ExerciseSlug = ""
			}
		}
	}
}
